File: aufgabe5/main.go
package main

import (
	"fmt"
	"math/rand"
)

const Processes = 10

type Process struct {
	pid    int
	active bool
}

func (p Process) String() string {
	var out = "inactive"
	if p.active {
		out = "active"
	}
	return fmt.Sprintf("p(%d, %s)", p.pid, out)
}

func (p Process) elect(processes []Process, sender int) bool {
	if !p.active {
		return false
	}
	if sender != -1 {
		fmt.Printf("OK-Nachricht von Prozess %d an Prozess %d gesendet\n", p.pid, sender)
	}

	var receivedOk = false
	for pid := p.pid + 1; pid < len(processes); pid++ {
		//for pid := len(processes) - 1; pid > p.pid; pid-- {
		fmt.Printf("ELECT-Nachricht von Prozess %d an Prozess %d gesendet\n", p.pid, pid)
		if processes[pid].elect(processes, p.pid) {
			receivedOk = true
			//break
		}
	}

	if !receivedOk {
		p.coord(processes)
	}
	return true
}

func (p Process) coord(processes []Process) {
	for pid := 0; pid < len(processes); pid++ {
		if processes[pid].active {
			fmt.Printf("COORD-Nachricht von Prozess %d an Prozess %d gesendet\n", p.pid, pid)
		}
	}
}

func main() {
	var processes = make([]Process, Processes)
	for i := 0; i < Processes; i++ {
		processes[i].pid = i
		processes[i].active = rand.Float32() < 1.5
	}

	fmt.Printf("%v\n", processes)

	var userDefinedProcessToKill int // rand.Int31n(Processes)
	fmt.Print("Process to kill: ")
	fmt.Scanf("%d\n", &userDefinedProcessToKill)
	var userDefinedProcessToStartElect int // rand.Int31n(Processes - 1)
	fmt.Print("Process for election: ")
	fmt.Scanf("%d\n", &userDefinedProcessToStartElect)

	processes[userDefinedProcessToKill].active = false

	fmt.Printf("%v\n", processes)

	// for _, process := range processes {
	// 	if process.active {
	// 		userDefinedProcessToStartElect = process.pid
	// 		break
	// 	}
	// }

	fmt.Printf("kill: %d elect: %d\n", userDefinedProcessToKill, userDefinedProcessToStartElect)

	processes[userDefinedProcessToStartElect].elect(processes, -1)
}

File: aufgabe5/main.c

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#define PROCESSES 10

typedef struct {
  int pid;
  bool active;
} process_t;

void process_coord(const process_t p, const process_t processes[PROCESSES]) {
  for (int pid = 0; pid < PROCESSES; pid++) {
    if (processes[pid].active) {
      printf("COORD-Nachricht von Prozess %d an Prozess %d gesendet\n", p.pid, pid);
    }
  }
}

bool process_elect(const process_t p, const process_t processes[PROCESSES], int sender) {
  if (!p.active) {
    return false;
  }
  if (sender != -1) {
    printf("OK-Nachricht von Prozess %d an Prozess %d gesendet\n", p.pid, sender);
  }
  bool receivedOk = false;
  for (int pid = p.pid + 1; pid < PROCESSES; pid++) {
    printf("ELECT-Nachricht von Prozess %d an Prozess %d gesendet\n", p.pid, pid);
    if (process_elect(processes[pid], processes, p.pid)) {
      receivedOk = true;
    }
  }

  if (!receivedOk) {
    process_coord(p, processes);
  }
  return true;
}

int main(int argc, char** argv) {
  process_t processes[PROCESSES];
  for (int i = 0; i < PROCESSES; i++) {
    processes[i].pid = i;
    processes[i].active = true;
  }

  if (argc < 3) {
    printf("%s <processToKill> <processToElect>\n", argv[0]);
    return EXIT_FAILURE;
  }

  const int userDefinedProcessToKill = atoi(argv[1]);
  const int userDefinedProcessToStartElect = atoi(argv[2]);

  if (userDefinedProcessToKill < 0 || userDefinedProcessToKill > PROCESSES) {
    fprintf(stderr, "process to kill must be in interval [0, %d]\n", PROCESSES);
    return EXIT_FAILURE;
  }
  if (userDefinedProcessToStartElect < 0 || userDefinedProcessToStartElect > PROCESSES) {
    fprintf(stderr, "process for elections must be in interval [0, %d]\n", PROCESSES);
    return EXIT_FAILURE;
  }

  processes[userDefinedProcessToKill].active = false;

  printf("kill: %d, elect: %d\n", userDefinedProcessToKill, userDefinedProcessToStartElect);

  process_elect(processes[userDefinedProcessToStartElect], processes, -1);

  return EXIT_SUCCESS;
}
